cmd/agent/app/snmpwalk.go

```diff
@@ -0,0 +1,316 @@
+				os.Exit(1)
+				return nil
+			}
+
+			// Privacy Protocol
+			switch strings.ToLower(privProt) {
+			case "":
+				privProtocol = gosnmp.NoPriv
+			case "des":
+				privProtocol = gosnmp.DES
+			case "aes":
+				privProtocol = gosnmp.AES
+			case "aes192":
+				privProtocol = gosnmp.AES192
+			case "aes192c":
+				privProtocol = gosnmp.AES192C
+			case "aes256":
+				privProtocol = gosnmp.AES256
+			case "aes256c":
+				privProtocol = gosnmp.AES256C
+			default:
+				fmt.Printf("Unsupported privacy protocol: %s \n", privProt)
+				cmd.Help() //nolint:errcheck
+				os.Exit(1)
+				return nil
+			}
+
+			// MsgFlags
+			switch strings.ToLower(securityLevel) {
+			case "":
+				msgFlags = gosnmp.NoAuthNoPriv
+				if privKey != "" {
+					msgFlags = gosnmp.AuthPriv
+				} else if authKey != "" {
+					msgFlags = gosnmp.AuthNoPriv
+				}
+
+			case "noauthnopriv":
+				msgFlags = gosnmp.NoAuthNoPriv
+			case "authpriv":
+				msgFlags = gosnmp.AuthPriv
+			case "authnopriv":
+				msgFlags = gosnmp.AuthNoPriv
+			default:
+				fmt.Printf("Unsupported security level: %s \n", securityLevel)
+				cmd.Help() //nolint:errcheck
+				os.Exit(1)
+				return nil
+			}
+		}
+		// Set SNMP parameters
+		snmp = gosnmp.GoSNMP{
+			Target:    deviceIP,
+			Port:      port,
+			Community: communityString,
+			Transport: "udp",
+			Version:   setVersion,
+			Timeout:   time.Duration(timeout * int(time.Second)),
+			Retries:   retries,
+			// v3
+			SecurityModel: gosnmp.UserSecurityModel,
+			ContextName:   snmpContext,
+			MsgFlags:      msgFlags,
+			SecurityParameters: &gosnmp.UsmSecurityParameters{
+				UserName:                 user,
+				AuthenticationProtocol:   authProtocol,
+				AuthenticationPassphrase: authKey,
+				PrivacyProtocol:          privProtocol,
+				PrivacyPassphrase:        privKey,
+			},
+		}
+		// Establish connection
+		err := snmp.Connect()
+		if err != nil {
+			fmt.Printf("Connect err: %v\n", err)
+			os.Exit(1)
+			return nil
+		}
+		defer snmp.Conn.Close()
+
+		// Perform a snmpwalk using Walk for all versions
+		err = snmp.Walk(oid, printValue)
+		if err != nil {
+			fmt.Printf("Walk Error: %v\n", err)
+			os.Exit(1)
+		}
+
+		return nil
+	},
+}
+
+func printValue(pdu gosnmp.SnmpPDU) error {
+
+	fmt.Printf("%s = ", pdu.Name)
+
+	switch pdu.Type {
+	case gosnmp.OctetString:
+		b := pdu.Value.([]byte)
+		fmt.Printf("STRING: %s\n", string(b))
+	case gosnmp.ObjectIdentifier:
+		fmt.Printf("OID: %s\n", pdu.Value)
+	case gosnmp.TimeTicks:
+		fmt.Print(pdu.Value, "\n")
+	case gosnmp.Counter32:
+		fmt.Printf("Counter 32: %d\n", pdu.Value.(uint))
+	case gosnmp.Counter64:
+		fmt.Printf("Counter 64: %d\n", pdu.Value.(uint64))
+	case gosnmp.Integer:
+		fmt.Printf("INTEGER: %d\n", pdu.Value.(int))
+	case gosnmp.Gauge32:
+		fmt.Printf("Gauge 32: %d\n", pdu.Value.(uint))
+	case gosnmp.IPAddress:
+		fmt.Printf("IpAddress: %s\n", pdu.Value.(string))
+	default:
+		fmt.Printf("TYPE %d: %d\n", pdu.Type, gosnmp.ToBigInt(pdu.Value))
+	}
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I would suggest 1/ concatenating first or event better using 2/ `strings.Builder` instead of calling fmt.Printf multiple time. Printf is much slower than concat and string builders.

For 1/, example:

```
    var value string
	switch pdu.Type {
	case gosnmp.OctetString:
		b := pdu.Value.([]byte)
		value = "STRING: " + string(b)  // using string concat is more efficient
	case gosnmp.ObjectIdentifier:
		value = "OID: " + pdu.Value
	case gosnmp.TimeTicks:
		[...]
	}
	fmt.Println(pdu.Name + " = " + value)
```

Reply:
While I agree generically on this, for an helper printer I think performance is not a priority. `fmt.Printf` is more readable IMO, so would keep this as is. 

`Printf is much slower than concat and string builders.` is it ms, ns, or seconds slower ? If it takes 10 seconds vs 1 second I would agree, if it takes 100ms vs 1 ms I don't think customer will really detect it.

Reply:
> I think performance is not a priority

I think it might matter the the device has huge number of OIDs e.g. (100k).

Let me try to quickly benchmark the different cases.

Reply:
Based on benchmark below against cisco-3850.snmprec (15493 OIDs), changing to string builder in this doesn't change the overall performance (possible because most of the time is spent on IO to retrieve OIDs).
So, nothing to fix here them.

Notice, I also did the benchmark for official snmpwalk, and we get very similar time.

```
Printf (current code)
Run 1: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.64s user 0.75s system 7% cpu 18.951 total
Run 2: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.66s user 0.76s system 7% cpu 18.595 total
Run 3: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.79s user 0.93s system 7% cpu 22.380 total

String builder
Run 1: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.64s user 0.74s system 7% cpu 18.200 total
Run 2: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.69s user 0.80s system 7% cpu 19.641 total
Run 3: ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850   0.67s user 0.76s system 7% cpu 18.800 total

Officiel snmpwalk:
Run 1: snmpwalk -O bentU -v 2c -c cisco-3850 127.0.0.1:1161  0.47s user 0.55s system 5% cpu 19.753 total
Run 2: snmpwalk -O bentU -v 2c -c cisco-3850 127.0.0.1:1161  0.46s user 0.53s system 5% cpu 18.611 total
Run 3: snmpwalk -O bentU -v 2c -c cisco-3850 127.0.0.1:1161  0.55s user 0.64s system 5% cpu 23.837 total
```